internal/controllers/replication/symphony.go

```diff
@@ -0,0 +1,223 @@
+package replication
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type symphonyController struct {
+	client client.Client
+}
+
+func NewSymphonyController(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apiv1.Symphony{}).
+		Owns(&apiv1.Composition{}).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "symphonyReplicationController")).
+		Complete(&symphonyController{
+			client: mgr.GetClient(),
+		})
+}
+
+func (c *symphonyController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	symph := &apiv1.Symphony{}
+	err := c.client.Get(ctx, req.NamespacedName, symph)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	logger = logger.WithValues("symphonyName", symph.Name, "symphonyNamespace", symph.Namespace)
+	ctx = logr.NewContext(ctx, logger)
+
+	existing := &apiv1.CompositionList{}
+	err = c.client.List(ctx, existing, client.MatchingFields{
+		manager.IdxCompositionsBySymphony: symph.Name,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing existing compositions: %w", err)
+	}
+
+	modified, err := c.syncStatus(ctx, symph, existing)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if modified {
```

Comment:
What's the reasoning behind using modified as a signal to stop reconciliation of these sub-operations? Is it worth adding comments?

Reply:
Passing down a bool is kind of strange looking, but it's the same behavior as other controllers: one write per tick of the loop.

This pattern guarantees that intermediate states will converge i.e. if the controller crashes between two writes, the behavior of the second write will be identical to if it hadn't crashed.

Reply:
Added a comment